bson/src/main/org/bson/codecs/pojo/RecordCreatorExecutable.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs.pojo;
+
+import org.bson.codecs.configuration.CodecConfigurationException;
+import org.bson.codecs.pojo.annotations.BsonId;
+import org.bson.codecs.pojo.annotations.BsonProperty;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.RecordComponent;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+
+final class RecordCreatorExecutable<T> implements CreatorExecutable<T> {
+    private final Class<T> clazz;
+    private final Constructor<T> constructor;
+    private final List<String> properties = new ArrayList<>();
+    private final Integer idPropertyIndex;
+    private final List<Class<?>> parameterTypes = new ArrayList<>();
+    private final List<Type> parameterGenericTypes = new ArrayList<>();
+
+    RecordCreatorExecutable(final Class<T> clazz, final Constructor<T> constructor) {
+        this.clazz = clazz;
+        this.constructor = constructor;
+        Integer idPropertyIndex = null;
+        parameterTypes.addAll(asList(constructor.getParameterTypes()));
+        parameterGenericTypes.addAll(asList(constructor.getGenericParameterTypes()));
+
+        RecordComponent[] recordComponents = clazz.getRecordComponents();
+        for (int i = 0; i < recordComponents.length; i++) {
+            RecordComponent recordComponent = recordComponents[i];
+            String name = recordComponent.getName();
+
+            boolean isId = name.equals("id") || name.equals("_id");
+            for (Annotation a : recordComponent.getAnnotations()) {
+                if (a.annotationType().equals(BsonId.class)) {
+                    isId = true;
+                    break;
+                }
+                if (a.annotationType().equals(BsonProperty.class)) {
```

Comment:
I'm not clear why a component annotated with BsonProperty can't also be the id.  Can you explain?  I may be misunderstanding given that AnnotationBsonPropertyIdRecord does exactly that.

Reply:
Thanks for asking this, there was a bug here... I had done this originally to get the test involving AnnotationBsonPropertyIdRecord to work. By specifying @BsonProperty("id") the client is actually saying they do not want that field to be stored as the _id (because the write name is now id, so it can't be). There's a corresponding POJO test that does the same thing which is where I got the test case from.

But you're right that a RecordComponent with a BsonProperty can still be the _id if the name is actually "_id." I fixed the code to check for that now and added two test cases to RecordRoundTripTest.